src/storage/dictionary_encoding_tile.cpp

```diff
@@ -0,0 +1,186 @@
+			LOG_INFO("encoding column %s", schema.GetColumn(i).column_name.c_str());
+			// to for tuple offset
+			for (oid_t to = 0; to < num_tuple_slots; to++) {
+				type::Value curr_old_val = tile->GetValue(to, i);
+				type::Value curr_val;
+				if (curr_old_val.GetTypeId() == type::TypeId::VARBINARY) {
+					curr_val = type::ValueFactory::GetVarbinaryValue((const unsigned char*) curr_old_val.GetData(),
+					    curr_old_val.GetLength(), true);
+				} else {
+					curr_val = type::ValueFactory::GetVarcharValue(curr_old_val.GetData(), curr_old_val.GetLength(),
+							true);
+				}
+				LOG_INFO("%s", curr_val.GetInfo().c_str());
+				// assume the idx take 1 byte
+				char idx_data[1];
+				if (dict.count(curr_val) == 0) {
+					LOG_INFO("Encoding!!!!!");
+					element_array.push_back(curr_val);
+					idx_data[0] = element_array.size() - 1;
+					dict.emplace(curr_val, idx_data[0]);
+				} else {
+					LOG_INFO("Already encoded!!!");
+					idx_data[0] = dict[curr_val];
+				}
+				// many constructor of Value is private, so use
+				// DeserializeFrom to construct the idx Value
+				type::Value idx_val(type::Value::DeserializeFrom(idx_data, type::TypeId::TINYINT, true));
+				new_data_vector[i].push_back(idx_val);
+			}
+		} else {
+			for (oid_t to = 0; to < num_tuple_slots; to++) {
+				new_data_vector[i].push_back(tile->GetValue(to, i));
+			}
+		}
+	}
+
+	// now we have all data in new_data_vector
+	// put new data into storage space
+
+	for (oid_t i = 0; i < column_count; i++) {
+		for (oid_t to = 0; to < num_tuple_slots; to++) {
+			SetValueFast(new_data_vector[i][to], to, schema.GetOffset(i), schema.IsInlined(i),
+					schema.GetLength(i));
+		}
+	}
+	is_dict_encoded = true;
+}
+
+Tile* DictEncodedTile::DictDecode() {
+	if (!is_dict_encoded) return nullptr;
+  LOG_INFO("dictionary decode, database_id: %d, table_id: %d, tile_group_id: %d"
+					", tile_id: %d", database_id, table_id, tile_group_id, tile_id);
+  // new data mean decoded data
+  std::vector<std::vector<type::Value>> new_data_vector(column_count);
+  for (oid_t i = 0; i < column_count; i++) {
+    if (dict_encoded_columns.count(i) > 0) {
+      LOG_INFO("decoding column %s", schema.GetColumn(i).column_name.c_str());
+      for (oid_t to = 0; to < num_tuple_slots; to++) {
+        type::Value curr_val = Tile::GetValue(to, i);
+        auto idx = curr_val.GetAs<uint8_t>();
+        type::Value actual_val = element_array[idx];
+        new_data_vector[i].push_back(actual_val);
+      }
+    } else {
+      for (oid_t to = 0; to < num_tuple_slots; to++) {
+        new_data_vector[i].push_back(GetValue(to, i));
+      }
+    }
+	}
+
+  auto *new_tile = new Tile(backend_type, tile_group_header, original_schema, tile_group, num_tuple_slots);
+  for (oid_t i = 0; i < column_count; i++) {
+		for (oid_t to = 0; to < num_tuple_slots; to++) {
+			new_tile->SetValueFast(new_data_vector[i][to], to, original_schema.GetOffset(i), original_schema.IsInlined(i),
```

Comment:
Wouldn't it be faster and use less memory to set the decoded Tile's values directly from the encoded tile's values? Why incur the cost of allocating, reading, and writing, to the separate "new_data_vector"?